JSON Descriptors framework

cristian-recoseanu opened the pull request:
- Add JSON descriptor source models
- Add full framework markdown with injected generated webIDL fragments

cristian-recoseanu committed:
JSON Descriptors framework

- Add JSON descriptor source models
- Add full framework markdown with injected generated webIDL fragments

cristian-recoseanu committed:
Try webidl highlighting

cristian-recoseanu committed:
Lint fixes

cristian-recoseanu committed:
Added paragraphs about extended attributes

cristian-recoseanu committed:
Remove as used in a feature set

cristian-recoseanu committed:
Refactor documentation

to integrate the newly published framework generated from JSON descriptors.

cristian-recoseanu committed:
Correct classID string form

jonathan-r-thorpe commented on docs/NcObject.md:
```diff
@@ -1,148 +1,52 @@
-| touchpoints       | sequence<NcTouchpoint>         | Yes          | sequence of touchpoint (see Touchpoints section for details)          |
-
-The `role` is a structural identifier which MUST be persisted across restarts. The role path of an element can be constructed from its role and the roles of all its parents. The role path can then be used to target a particular section of the control model tree or to retrieve the object ids again after a restart.
+The `role` is a structural identifier which MUST be persisted across restarts. asdasda
```
Spurious "asdasda" at the end of the line

garethsb replied:
Shh. Pretty sure that's a magic word like xyzzy 🧙‍♂️

jonathan-r-thorpe commented on models/classes/1.1.json:
```diff
@@ -0,0 +1,293 @@
+      "isDeprecated": false
+    },
+    {
+      "description": "this block's ports",
```
Inconsistent capitalization of the description fields. Should the first letter not be capitalized here?

cristian-recoseanu replied:
Thanks. I will fix all the inconsistent capitalizations.

jonathan-r-thorpe commented on models/classes/1.1.json:
```diff
@@ -0,0 +1,293 @@
+      "isDeprecated": false
+    },
+    {
+      "description": "this block's signal paths",
```
Capitalization as above

jonathan-r-thorpe commented on models/classes/1.1.json:
```diff
@@ -0,0 +1,293 @@
+  ],
+  "methods": [
+    {
+      "description": "gets descriptors of members of the block",
```
Capitalization as above

jonathan-r-thorpe commented on models/classes/1.1.json:
```diff
@@ -0,0 +1,293 @@
+      "isDeprecated": false
+    },
+    {
+      "description": "finds member(s) by path",
```
Capitalization as above

jonathan-r-thorpe commented on models/classes/1.1.json:
```diff
@@ -0,0 +1,293 @@
+      "resultDatatype": "NcMethodResultBlockMemberDescriptors",
+      "parameters": [
+        {
+          "description": "relative path to search for (MUST not include the role of the block targeted by oid)",
```
Capitalization as above

jonathan-r-thorpe commented on models/classes/1.1.json:
```diff
@@ -0,0 +1,293 @@
+      "isDeprecated": false
+    },
+    {
+      "description": "finds members with given role name or fragment",
```
Capitalization as above

jonathan-r-thorpe commented on models/classes/1.1.json:
```diff
@@ -0,0 +1,293 @@
+      "resultDatatype": "NcMethodResultBlockMemberDescriptors",
+      "parameters": [
+        {
+          "description": "role text to search for",
```
Capitalization as above

jonathan-r-thorpe commented on models/classes/1.1.json:
```diff
@@ -0,0 +1,293 @@
+          "isSequence": false
+        },
+        {
+          "description": "signals if the comparison should be case sensitive",
```
Capitalization as above

jonathan-r-thorpe commented on models/classes/1.1.json:
```diff
@@ -0,0 +1,293 @@
+      "isDeprecated": false
+    },
+    {
+      "description": "finds members with given class id",
```
Capitalization as above

jonathan-r-thorpe commented on models/classes/1.1.json:
```diff
@@ -0,0 +1,293 @@
+      "resultDatatype": "NcMethodResultBlockMemberDescriptors",
+      "parameters": [
+        {
+          "description": "class id to search for",
```
Capitalization as above

jonathan-r-thorpe commented on models/classes/1.3.1.json:
```diff
@@ -0,0 +1,164 @@
+  "fixedRole": "DeviceManager",
+  "properties": [
+    {
+      "description": "Version of nc this dev uses",
```
Should "nc" be nca or ncp ?  This dev thinks you should device rather than dev :)

cristian-recoseanu replied:
Not NCP as that is covered by IS-12 and the version in the urn published in the controls array of a device.
This is possibly the version of MS-05-02 your models are working with. So maybe we should just say that?

cristian-recoseanu committed:
Remove random sequence of characters

jonathan-r-thorpe commented on models/classes/1.3.2.json:
```diff
@@ -0,0 +1,97 @@
+          "isSequence": false
+        },
+        {
+          "description": "if set the descriptor would contain all inherited elements",
```
Capitalization as above

jonathan-r-thorpe commented on models/classes/1.3.2.json:
```diff
@@ -0,0 +1,97 @@
+          "isSequence": false
+        },
+        {
+          "description": "if set the descriptor would contain all inherited elements",
```
Capitalization as above

jonathan-r-thorpe commented on models/classes/1.json:
```diff
@@ -0,0 +1,292 @@
+      "isDeprecated": false
+    },
+    {
+      "description": "role of obj in containing block",
```
Capitalization as above

jonathan-r-thorpe commented on models/datatypes/NcOrganizationId.json:
```diff
@@ -0,0 +1,7 @@
+{
+    "description": "Unique 24-bit organization ID",
```
Consistency with capitalization of ID/id. Lower case for "Object id" so should be lower case here for consistency?

jonathan-r-thorpe commented on models/datatypes/NcParameterConstraintsNumber.json:
```diff
@@ -0,0 +1,29 @@
+  "type": 2,
+  "fields": [
+    {
+      "description": "optional maximum",
```
Capitalization as above

jonathan-r-thorpe commented on models/datatypes/NcParameterConstraintsNumber.json:
```diff
@@ -0,0 +1,29 @@
+      "isSequence": false
+    },
+    {
+      "description": "optional minimum",
```
Capitalization as above

jonathan-r-thorpe commented on models/datatypes/NcParameterConstraintsNumber.json:
```diff
@@ -0,0 +1,29 @@
+      "isSequence": false
+    },
+    {
+      "description": "optional step",
```
Capitalization as above

jonathan-r-thorpe commented on models/datatypes/NcParameterConstraintsString.json:
```diff
@@ -0,0 +1,22 @@
+  "type": 2,
+  "fields": [
+    {
+      "description": "maximum characters allowed",
```
Capitalization as above

jonathan-r-thorpe commented on models/datatypes/NcParameterConstraintsString.json:
```diff
@@ -0,0 +1,22 @@
+      "isSequence": false
+    },
+    {
+      "description": "regex pattern",
```
Capitalization as above

jonathan-r-thorpe commented on models/datatypes/NcPropertyConstraints.json:
```diff
@@ -0,0 +1,29 @@
+  "type": 2,
+  "fields": [
+    {
+      "description": "relative path to member (null means current member)",
```
Capitalization as above

jonathan-r-thorpe commented on models/datatypes/NcPropertyConstraints.json:
```diff
@@ -0,0 +1,29 @@
+      "isSequence": false
+    },
+    {
+      "description": "ID of property being constrained",
```
Difficult one as it's an inconsistent capitalization of ID, but lower case id would mean the sentence starts with a lower case character

jonathan-r-thorpe commented on models/datatypes/NcPropertyConstraints.json:
```diff
@@ -0,0 +1,29 @@
+      "isSequence": false
+    },
+    {
+      "description": "optional default value",
```
Capitalization as above

jonathan-r-thorpe commented on models/datatypes/NcPropertyConstraintsNumber.json:
```diff
@@ -0,0 +1,29 @@
+  "type": 2,
+  "fields": [
+    {
+      "description": "optional maximum",
```
Capitalization as above

jonathan-r-thorpe commented on models/datatypes/NcPropertyConstraintsNumber.json:
```diff
@@ -0,0 +1,29 @@
+      "isSequence": false
+    },
+    {
+      "description": "optional minimum",
```
Capitalization as above

jonathan-r-thorpe commented on models/datatypes/NcPropertyConstraintsNumber.json:
```diff
@@ -0,0 +1,29 @@
+      "isSequence": false
+    },
+    {
+      "description": "optional step",
```
Capitalization as above

jonathan-r-thorpe commented on models/datatypes/NcPropertyConstraintsString.json:
```diff
@@ -0,0 +1,22 @@
+  "type": 2,
+  "fields": [
+    {
+      "description": "maximum characters allowed",
```
Capitalization as above

jonathan-r-thorpe commented on models/datatypes/NcPropertyConstraintsString.json:
```diff
@@ -0,0 +1,22 @@
+      "isSequence": false
+    },
+    {
+      "description": "regex pattern",
```
Capitalization as above

jonathan-r-thorpe commented on models/datatypes/NcTouchpointResourceNmosChannelMapping.json:
```diff
@@ -0,0 +1,15 @@
+    "type": 2,
+    "fields": [
+        {
+            "description": "IS-08 Audio Channel Mapping input or output ID",
```
Inconsistent capitalization of ID. Should be lower case?

peterbrightwell committed:
Use typescript formatting for webidl in markdown

jonathan-r-thorpe commented on docs/Blocks.md:
````diff
@@ -1,78 +1,27 @@
-    attribute sequence<NcPropertyConstraints>? constraints  // Constraints on this member or, for a block, its members.
-};
-```
+Blocks enable device tree discovery by offering the descriptors of their contained members in the `members` property which holds a collection of type [NcBlockMemberDescriptor](Framework.md#ncblockmemberdescriptor).
````
Perhaps normative language is needed here:

```suggestion
Blocks MUST enable device tree discovery by offering the descriptors of their contained members in the `members` property which holds a collection of type [NcBlockMemberDescriptor](Framework.md#ncblockmemberdescriptor).
```

cristian-recoseanu replied:
Isn't this covered by the umbrela conformance language at the top of the Framework document which says if you are using the models you must conform to their specification?
https://github.com/AMWA-TV/ms-05-02/blob/CR-json-descriptors/docs/Framework.md

jonathan-r-thorpe replied:
Yes, fair point.

On a separate point with reference to #42 which term is going to be the preferred: Device Model or Device Tree?  Given that Device Model is the term defined in [MS-05-01](https://github.com/AMWA-TV/ms-05-01/blob/e9f03c860030452ab08bc0123f240b66c1fb6ada/docs/Device%20Model.md#device-model) should we not be using that instead?

jonathan-r-thorpe commented on docs/Blocks.md:
```diff
@@ -1,78 +1,27 @@
 
 ## Search methods
 
 All blocks also offer some search methods for convenience:
```
Perhaps firm this up with normative language (MUST/MAY)?

cristian-recoseanu replied:
Isn't this covered by the umbrela conformance language at the top of the Framework document which says if you are using the models you must conform to their specification?
https://github.com/AMWA-TV/ms-05-02/blob/CR-json-descriptors/docs/Framework.md

jonathan-r-thorpe commented on docs/Feature sets.md:
```diff
@@ -1,42 +1,7 @@
-A derived receiver monitor for SMPT ST 2022-7 type receivers also exists as `NcReceiverMonitorProtected`.
-
-This adds a further `NcBoolean` status `signalProtectionStatus` to indicate if the signal protection is enabled.
+Where a device employs functionality covered by a feature set it MUST comply with the model definitions specified in the particular feature set.
```
```suggestion
Where a device employs functionality covered by a feature set it MUST comply with the model definitions specified for that feature set in the Control feature sets register.
```

jonathan-r-thorpe commented on docs/Managers.md:
```diff
@@ -1,6 +1,7 @@
 # Managers
 
-Managers are special classes which collate information which pertains to the entire device. Each manager class applies to a specific functional context. All managers must inherit from `NcManager`.
+Managers are singleton (can only be instantiated once) classes which collate information which pertains to the entire device. Each manager class applies to a specific functional context.  
```
Two "which"s in quick succession

```suggestion
Managers are singleton (can only be instantiated once) classes which collate information that pertains to the entire device. Each manager class applies to a specific functional context.  
```

cristian-recoseanu committed:
Update docs/Feature sets.md

Wording suggestion inclusion.

Co-authored-by: jonathan-r-thorpe <64410119+jonathan-r-thorpe@users.noreply.github.com>

jonathan-r-thorpe commented on docs/Managers.md:
```diff
@@ -1,6 +1,7 @@
 
-Managers are special classes which collate information which pertains to the entire device. Each manager class applies to a specific functional context. All managers must inherit from `NcManager`.
+Managers are singleton (can only be instantiated once) classes which collate information which pertains to the entire device. Each manager class applies to a specific functional context.  
+All managers must inherit from [NcManager](Framework.md#ncmanager).
```
```suggestion
All managers MUST inherit from [NcManager](Framework.md#ncmanager).
```

cristian-recoseanu committed:
Update docs/Managers.md

Inclusion of suggestion from review.

Co-authored-by: jonathan-r-thorpe <64410119+jonathan-r-thorpe@users.noreply.github.com>

cristian-recoseanu committed:
Update docs/Managers.md

Include suggestion from review.

Co-authored-by: jonathan-r-thorpe <64410119+jonathan-r-thorpe@users.noreply.github.com>

jonathan-r-thorpe commented on docs/Managers.md:
```diff
@@ -16,114 +17,23 @@ Example:
+* datatypes (lists all data type descriptors in the device using the [NcDatatypeDescriptor](Framework.md#ncdatatypedescriptor) type - descriptors do not contain inherited elements)
 
-The descriptor for an individual data type may be retrieved using the `GetDatatype` method (`[element("3m2")]`) and passing the name (type `NcName`) and includeInherited (if all inherited elements should be included - type `NcBoolean`) as arguments. The method has a response of type `NcMethodResultDatatypeDescriptor`.
+The descriptor for an individual control class may be retrieved using the `GetControlClass` method (`[element("3m1")]`) and passing the identity (type [NcClassId](Framework.md#ncclassid)) and includeInherited (if all inherited elements should be included - type [NcBoolean](Framework.md#primitives)) as arguments. The method has a response of type [NcMethodResultClassDescriptor](Framework.md#ncmeth ...
```
Is the use of "may" here normative? Perhaps substitute "is" to avoid confusion?

cristian-recoseanu replied:
I think "can" may be a better choice? The idea here is that the descriptors can also be retrieved from the ClassManager properties if you don't require inheritance information so this then becomes an option.

jonathan-r-thorpe replied:
Yes, "can" works well.

jonathan-r-thorpe commented on docs/Managers.md:
````diff
@@ -16,114 +17,23 @@ Example:
-    attribute NcDatatypeDescriptor value;
-};
-```
+The descriptor for an individual data type may be retrieved using the `GetDatatype` method (`[element("3m2")]`) and passing the name (type [NcName](Framework.md#ncname)) and includeInherited (if all inherited elements should be included - type [NcBoolean](Framework.md#primitives)) as arguments. The method has a response of type [NcMethodResultDatatypeDescriptor](Framework.md#ncmethodresultdatatyp ...
````
Same as above: may => is ?

cristian-recoseanu replied:
Same as above possibly?

jonathan-r-thorpe replied:
Yep, I think so.

cristian-recoseanu committed:
Updates as per review comments

- fixed capitalization issues for descriptors
- used device model term instead of device tree
- swapped "may" with "can" in the ClassManager section

cristian-recoseanu commented:
Also reviewed on the call on the 9th of May 2023.
